internal/component/prometheus/remote/queue/filequeue/filequeue.go

```diff
@@ -0,0 +1,182 @@
+		if err != nil {
+			continue
+		}
+		ids[i] = id
+	}
+	sort.Ints(ids)
+	var currentIndex int
+	if len(ids) > 0 {
+		currentIndex = ids[len(ids)-1]
+	}
+	q := &queue{
+		directory:      directory,
+		maxIndex:       currentIndex,
+		logger:         logger,
+		out:            out,
+		files:          actor.NewMailbox[types.Data](),
+		existingsFiles: make([]string, 0),
+	}
+
+	// Push the files that currently exist to the channel.
+	for _, id := range ids {
+		name := filepath.Join(directory, fmt.Sprintf("%d.committed", id))
+		q.existingsFiles = append(q.existingsFiles, name)
+	}
+	return q, nil
+}
+
+func (q *queue) Start() {
+	q.self = actor.Combine(actor.New(q), q.files).Build()
+	q.self.Start()
+
+}
+
+func (q *queue) Stop() {
+	q.self.Stop()
+}
+
+func (q *queue) Store(ctx context.Context, meta map[string]string, data []byte) error {
+	return q.files.Send(ctx, types.Data{
+		Meta: meta,
+		Data: data,
+	})
+}
+
+// get returns the data of the file or an error if something wrong went on.
+func get(name string) (map[string]string, []byte, error) {
+	defer deleteFile(name)
+	buf, err := readFile(name)
+	if err != nil {
+		return nil, nil, err
+	}
+	r := &Record{}
+	_, err = r.UnmarshalMsg(buf)
+	if err != nil {
+		return nil, nil, err
+	}
+	return r.Meta, r.Data, nil
+}
+
+func (q *queue) DoWork(ctx actor.Context) actor.WorkerStatus {
+	// Queue up our existing items.
+	for _, name := range q.existingsFiles {
+		q.out(ctx, types.DataHandle{
+			Name: name,
+			Get: func() (map[string]string, []byte, error) {
+				return get(name)
+			},
+		})
+	}
+	// We only want to process existing files once.
+	q.existingsFiles = nil
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case item, ok := <-q.files.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		name, err := q.add(item.Meta, item.Data)
+		if err != nil {
+			level.Error(q.logger).Log("msg", "error adding item", "err", err)
+			return actor.WorkerContinue
+		}
+		q.out(ctx, types.DataHandle{
+			Name: name,
+			Get: func() (map[string]string, []byte, error) {
+				return get(name)
+			},
+		})
+		return actor.WorkerContinue
+	}
+}
+
+// Add a committed file to the queue.
+func (q *queue) add(meta map[string]string, data []byte) (string, error) {
+	if meta == nil {
+		meta = make(map[string]string)
+	}
+	q.maxIndex++
+	name := filepath.Join(q.directory, fmt.Sprintf("%d.committed", q.maxIndex))
+	// record wraps the data and metadata in one. This allows the consumer to take action based on the map.
+	r := &Record{
+		Meta: meta,
+		Data: data,
+	}
+	// TODO @mattdurham reuse a buffer here.
+	rBuf, err := r.MarshalMsg(nil)
+	if err != nil {
+		return "", err
+	}
+	err = q.writeFile(name, rBuf)
+	if err != nil {
+		return "", err
+	}
+	return name, err
+}
+
+func (q *queue) writeFile(name string, data []byte) error {
+	return os.WriteFile(name, data, 0644)
+}
+
+func deleteFile(name string) {
+	_ = os.Remove(name)
```

Comment:
probably will never happen, but I'd log on warn level the error... 
the number of times I had failed disk produce no useful error messages ...